src/nvim/eval.c

```diff
@@ -9501,24 +9501,35 @@ static void f_getchar(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   varnumber_T n;
   int error = FALSE;
 
-  /* Position the cursor.  Needed after a message that ends in a space. */
-  ui_cursor_goto(msg_row, msg_col);
-
   ++no_mapping;
   ++allow_keys;
   for (;; ) {
-    if (argvars[0].v_type == VAR_UNKNOWN)
-      /* getchar(): blocking wait. */
+    // Position the cursor.  Needed after a message that ends in a space,
+    // or if event processing caused a redraw.
+    ui_cursor_goto(msg_row, msg_col);
+
+    if (argvars[0].v_type == VAR_UNKNOWN) {
+      // getchar(): blocking wait.
+      if (!(char_avail() || using_script() || input_available())) {
+        input_enable_events();
+        (void)os_inchar(NULL, 0, -1, 0);
+        input_disable_events();
+        if (!queue_empty(main_loop.events)) {
+          queue_process_events(main_loop.events);
```

Comment:
`event/queue.h` was renamed to `multiqueue.h`, so this is now `multiqueue_process_events`, `multiqueue_empty`, etc.